tutorials/review.adoc

```diff
@@ -0,0 +1,881 @@
+----
+With this command you trigger the metanorma toolchain to:
+
+* Read the AsciiDoc input
+* Convert it into XML
+* Check the XML against the document model (XML schema)
+* Create HTML, PDF, and DOC output
+
+[[exercise-3-2]]
+include::exercises.adoc[tag=exercise-3-2]
+
+
+Now that you are familiar with the simplest way of creating a Metanorma document, let's look at different ways to compile documents in the next lesson.
+>> Start next lesson
+
+== Publishing a Metanorma Document
+tag::[learningobjectives-4]
+
+Once you are happy with your document, the next step is to generate an output to send your SDO for publishing. Since Metanorma poses some validity criteria for your document, it is normal to face some errors. Don’t worry, we will go through some common errors and teach you how to resolve them. 
+In this lesson you will learn:
+
+* How to generate PDF, HTML, and Word output using the command line
+* How to troubleshoot common errors 
+
+end::[learningobjectives-4]
+
+>> Start next lesson
+
+=== Compiling a Metanorma Document
+
+To successfully compile a Metanorma document, the toolchain looks at:
+
+* Metadata: The title, document flavor, and document type.
+* Syntax: Is the entered AsciiDoc code correct?
+* Dependencies: Are all tools needed to compile the document installed?
+// Anything else?
+
+To compile a Metanorma document:
+
+. On the command line, go to the folder where the document you want to compile is located. 
+. Enter the following command: +
+[source, shell]
+----
+metanorma document.adoc
+----
+With this command you trigger the metanorma toolchain to:
+
+* Read the AsciiDoc input
+* Convert it into XML
+* Check the XML against the document model (XML schema)
+* Create HTML, PDF, and DOC output
+
+You can also manipulate the way Metanorma compiles a document by setting flags. Flags are appended to the build command, like this:  
+
+.Example of a build command with a flag
+[source, shell]
+----
+metanorma document.adoc -x html
+----
+
+You can use the following flags to manipulate the building process:
+
+* `-t`: Sets the flavor of the document +
+NOTE: Either define the flavor in the metadata using `mn-document-class` or use `-t flavor`
+* `-x`: Sets the output format(s) of the document 
+* `-o`: Enter a path that the output should be saved to. 
+
+To see the full list of possible build commands, open the metanorma help on the command line.
+
+[source, shell]
+----
+metanorma help compile
+----
+
+>> Start next lesson
+
+=== Troubleshooting 
+
+There can be many reasons why you can’t compile a Metanorma AsciiDoc document into the final output, but you can easily fix them when you know what typical errors occur and how to fix them. 
+Metanorma tells you what’s wrong in the terminal and stores all errors in an Error file `.err` where your original document is stored. 
+
+////
+Your input needed!
+What are the most common errors when compiling a document and how they are caused?
+How to resolve the errors?
+
```

Comment:
Generalities from me:


* Break document into bits and compile them, so you can narrow down the error location. 
* Recompile the document incrementally as you go while authoring it, so you can detect errors early and pin them down to recent additions. 
* Document outright crashes, because of error in markup. Validation errors start looking strange after a particular line number. Bug crash reports that execution failed in processing  a particular kind of content. 
* Document looks odd when rendered. Look through the document to see how; for example, incorrect number of `==` before a section title will result in wrong numbering of clauses. Sections can outright disappear if they are not where Metanorma expects to find them.
* Typos in references will result in XML crashes mentioning `IDREF` (the XML for a cross-reference within an XML file). Mentions of anchors have to be letter-perfect.
* Details of XML errors can be ignored in a first pass, and Metanorma is more pedantic about structure than SDOs usually are. However Style errors, expressed as text (e.g. about how to express references to normative text, or how to represent numbers), do correspond to rules set by the SDOs.

Reply:
@kickoke Sorry for waiting for my review. I still have a few things left to check (and try to remind myself about the issues I've had so far), so I'll add further comments tomorrow, if any.

For now, in addition to @opoudjis's comments, I'd like to add following:

- Make sure all gems are installed and up-to-date. In case gems are not installed, you will probably get an error such as: "Run `bundle install` to install missing gems." which will guide you to their installation, but it is also good to update them every once in a while even without getting such an error.

- It is a good practice to divide document into sections. You can then compile each section separately, or compile several of them in order to locate the error easier. For example, if you assume there could be some error in Foreword section and therefore want to exclude it from being rendered, you can easily add a delimiter for inline comment before the line which includes that section: `//include::sections/00-foreword.adoc[]`.

- In case there is some unrecognized cross-reference, make sure you are always using the same keyword for one reference.

- If document looks odd from one point onwards, you should probably check if you have added starting and ending delimiters for all block elements (such as table block, example block, source block, etc.) properly. If not, you will most probably receive some error message which points to the issue. For example: "unterminated example block", "table missing leading separator" or "unterminated table block".
In case when document contains source text from some point onwards, many errors and warnings can be shown in the terminal, especially regarding references and cross-references. Therefore, it is always good to see the actual output document before starting troubleshooting based on to the errors shown in the terminal.

- There are some mandatory document attributes for each flavor. Make sure you use them in your .adoc input. However, if you forget to add some of them, or if you use some unrecognized value for any of them, Metanorma will either automatically correct that error and inform you about it, or it will point which attribute should be added or changed.

- In case some figure cannot be loaded, make sure you added it to the correct path, indicated by `:imagesdir:` document attribute, or in the command for adding the figure itself. Also make sure that you wrote a correct extension.

- If you see some index term showing up multiple times when it should appear only once, you should check the parentheses used in index entries for that term.

Reply:
Thank you @anermina and @opoudjis for the input. 
I'll craft a text and let you know once it's ready for review.